File: skills/udf-gen-test/templates/java/src/test/java/com/udf/CudfComparisonTest.java
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

public class CudfComparisonTest {

    private static SparkSession spark;
    private static ClassLoader origContextClassLoader;

    @BeforeClass
    public static void setUp() {
        origContextClassLoader = TestUtils.installMutableClassLoader();
        spark = SparkSession.builder()
            .appName("UDF vs. RapidsUDF Comparison Test")
            .master("local[*]")
            .config("spark.plugins", "com.nvidia.spark.SQLPlugin")
            .config("spark.rapids.memory.gpu.pool", "NONE")
            .config("spark.rapids.sql.explain", "NONE")
            .enableHiveSupport()
            .getOrCreate();
    }

    @AfterClass
    public static void tearDown() {
        if (spark != null) spark.stop();
        if (origContextClassLoader != null) {
            Thread.currentThread().setContextClassLoader(origContextClassLoader);
        }
    }

    /** TODO: Register the RapidsUDF with Spark. */
    public static void registerRapidsUDF(SparkSession spark, String udfName) { }

    @Test
    public void testCpuVsRapidsUDF() {
        Dataset<Row> testDF = UnitTest.createTestData(spark).repartition(1);

        // Run CPU UDF
        UnitTest.registerUDF(spark, "placeholder_udf_name");
        Dataset<Row> cpuResultDF = UnitTest.executeUDF(
            spark, "placeholder_udf_name", testDF);
        UnitTest.verifyUDFResults(cpuResultDF, testDF);

        // Run RapidsUDF
        registerRapidsUDF(spark, "placeholder_rapids_udf_name");
        Dataset<Row> gpuResultDF = UnitTest.executeUDF(
            spark, "placeholder_rapids_udf_name", testDF);
        UnitTest.verifyUDFResults(gpuResultDF, testDF);

        // Compare
        TestUtils.assertDataFrameEquals(gpuResultDF, cpuResultDF);
    }

    /**
     * TODO: If UnitTest adds extra @Test methods beyond the main result checks,
     * add corresponding comparison tests here. Each case should run the same input
     * through the CPU UDF and the RapidsUDF, apply equivalent assertions to both
     * outputs, and compare the RapidsUDF output against the CPU output.
     */
}

File: skills/udf-gen-test/templates/scala/src/test/scala/com/udf/CudfComparisonTest.scala
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf

import org.apache.spark.sql.SparkSession
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.BeforeAndAfterAll

class CudfComparisonTest extends AnyFunSuite with BeforeAndAfterAll {

  var spark: SparkSession = _

  override def beforeAll(): Unit = {
    spark = SparkSession.builder()
      .appName("UDF vs. RapidsUDF Comparison Test")
      .master("local[*]")
      .config("spark.plugins", "com.nvidia.spark.SQLPlugin")
      .config("spark.rapids.memory.gpu.pool", "NONE")
      .config("spark.rapids.sql.explain", "NONE")
      .getOrCreate()
  }

  override def afterAll(): Unit = {
    if (spark != null) spark.stop()
  }

  /** TODO: Register the RapidsUDF with Spark. */
  def registerRapidsUDF(spark: SparkSession, udfName: String): Unit = ???

  test("UDF vs RapidsUDF") {
    val testDF = UnitTest.createTestData(spark).repartition(1)

    // Run CPU UDF
    UnitTest.registerUDF(spark, "placeholder_udf_name")
    val cpuResultDF = UnitTest.executeUDF(spark, "placeholder_udf_name", testDF)
    UnitTest.verifyUDFResults(cpuResultDF, testDF)

    // Run RapidsUDF
    registerRapidsUDF(spark, "placeholder_rapids_udf_name")
    val gpuResultDF = UnitTest.executeUDF(spark, "placeholder_rapids_udf_name", testDF)
    UnitTest.verifyUDFResults(gpuResultDF, testDF)

    // Compare
    TestUtils.assertDataFrameEquals(actual = gpuResultDF, expected = cpuResultDF)
  }

  /**
   * TODO: If UnitTest adds extra tests beyond the main result checks, add
   * corresponding comparison tests here. Each case should run the same input
   * through the CPU UDF and the RapidsUDF, apply equivalent assertions to both
   * outputs, and compare the RapidsUDF output against the CPU output.
   */
}
